### ISSUE TITLE
How to get the controller's name and action?

### ISSUE THREAD

**Issue by xpbliss** [colegion/goal#37]
How to get the controller's name and action?
How to get all the names and actions but "Before,After,Initially,Finnally"
I want to intercept some actions for authority.

**Comment by alkchr** [comment]
To get current controller and action names you can use `Initially` (or `Finally`) method:
```go
type App struct {
	*Controllers
	contr, action string
}

func (c *App) Initially(w http.ResponseWriter, r *http.Request, a []string) bool {
	// The third argument "a" is guaranteed to contain 2 elements:
	// controller name + action name.
	c.contr = a[0]
	c.action = a[1]
}

func (c *App) Index() http.Handler {
	// Now you know that current controller and action names are:
	println(c.contr) // App
	println(c.action) // Index
}

func (c *App) Profile() http.Handler {
	println(c.contr) // App
	println(c.action) // Profile
}
```
[Here](https://github.com/colegion/goal/blob/master/internal/skeleton/assets/handlers/app.go#L90) is a fragment of auto generated code that demonstrates where that parameter is initialized.

There is currently no way to get all action names without `reflect` package. Though, as I have mentioned earlier in https://github.com/colegion/goal/issues/36#issuecomment-174437652 it is not difficult to add that functionality to the `goal generate handlers`. I just want to understand the benefits of this first.

**Comment by xpbliss** [comment]
Get all the actions,in order to manage the action permission. 
assign the different user have the different action permissions.

**Comment by alkchr** [comment]
@xpbliss I updated the signature of controller's `Init` function, now it takes `url.Values` argument as input:
```go
import (
	"log"
	"net/url"
)

type App struct {
}

...

// NOTE: now it takes url.Values, before it was just Init().
func Init(ctx url.Values) {
	log.Println(ctx)
	// ctx := map[string][]string {
	// 	"App":      {"Index", "About", "Contact"},
	// 	"Profiles": {"Index", "View"},
	// 	"...":      {...},
	// }
	// ctx includes controllers with real actions only,
	// not "magic actions" such as Before or After.
}
```
You'll have to update your `Init` functions and upgrade `contrib` and `goal` packages:
```bash
go get -u github.com/colegion/contrib/...
go get -u github.com/colegion/goal
```

**Comment by xpbliss** [comment]
good! it  is resolved!

**Comment by xpbliss** [comment]
after that ,the" pongoal example" sample  can't find the rendered page

**Comment by alkchr** [comment]
@xpbliss Oh, right I had to update its `Init` function, too. Now it's done.

**Comment by xpbliss** [comment]
Maybe it is an ugly way for the getting all the actions. every controller maybe add `Init()` func?
 a common variable to store the actions in the controlles may be more better.
such as Controllers.Actions,

**Comment by alkchr** [comment]
Where that global variable should be located? In the generated `./assets/``handlers` package? It is actually [there](https://github.com/colegion/goal/blob/master/internal/skeleton/assets/handlers/app.go#L22). But you cannot import `handlers` package from your controllers as `handlers` already [import](https://github.com/colegion/goal/blob/master/internal/skeleton/assets/handlers/app.go#L9) controllers. And import cycle is not allowed in Go.

**Comment by xpbliss** [comment]
```
type Controllers struct {
	*requests.Requests
	*pongoal.Pongo2
	*sessions.Sessions
	Actions []string   //?
}
```

**Comment by alkchr** [comment]
A list of actions is static, it is defined once during code generation stage and that's it. As opposed to that Controller is allocated and initialized with every request again and again. So, it's not quite clear to me why that information must be within a controller. Moreover, apart from "magic methods" and "magic actions" we have already had, what you've showed would require us to introduce a new concept of "magic fields". And that's a lot of unnecessary complication.

Though you still can achieve that behaviour if you want without changes to `goal` repo. Update your `./controllers/init.go`:
```go
package controllers

import (
	"net/http"
	"net/url"

	...
)

var data url.Values

type Controllers struct {
	...
	Actions []string
}

...

// Initially sets the list of actions supported by current controller.
func (c *Controllers) Initially(w http.ResponseWriter, r *http.Request, a []string) bool {
	c.Actions = data[a[0]]
	return false
}

// Init saves information about all supported actions to the local variable.
func Init(ctx url.Values) {
	data = ctx
}
```
That's it. Now all your actions of `controllers` package can access `c.Actions`.

**Comment by xpbliss** [comment]
in pongoal.go,  i don't understand the affect  of the `func Init(_ url.Values) `

**Comment by alkchr** [comment]
#### Short answer
@xpbliss Simply saying, `Init` is somehow similar to the following construction in Revel:
```go
func init() {
	revel.OnAppStart(func() {
	})
}
```
But instead of a global variable we use code generation to accomplish the same behaviour with `Init(url.Values)`. The functions started the same time as `main()` just once to initialize controllers.

#### Detailed answer
During code generation all `Init(url.Values)` functions of controller packages are collected together and a [single one](https://github.com/colegion/goal/blob/master/internal/skeleton/assets/handlers/app.go#L115) is generated in `handlers` package:
```go
func Init() {
	initApp()
	initControllers()
	initErrors()
}
```
I.e. if we have the following `Controllers` type:
```go
package controllers
type Controllers struct {
	*db.DBController
	*sess.SessionController
}

func Init(_ url.Values) {}
```
The automatically generated (`handlers`).`Init` will look as follows:
```go
func Init() {
	db.Init(context)
	sess.Init(context)
	controllers.Init(context)
}
```
That generated [`Init` function](https://github.com/colegion/goal/blob/master/internal/skeleton/assets/handlers/app.go#L115) may be used whenever you want. But by default it is called [in the `main()`](https://github.com/colegion/goal/blob/master/internal/skeleton/main.go#L11) of your app (see [`server` package](https://github.com/colegion/goal/blob/master/internal/skeleton/server/server.go)).

#### pongoal
`pongoal`'s `Init` is similar to Revel's code:
```go
func init() {
	revel.OnAppStart(func() {
		// Scan and load all templates from `./app/views`, parse them...
		...

		// Make them available for use from my actions.
		revel.Templates = ...
	})
}
```

#### Related
Also related: [request life cycle](https://colegion.github.io/goal/manual/handlers/controllers.html#request-life-cycle).

**Comment by xpbliss** [comment]
Why the Initially and the finally do not execute?
```
	var err error
	engine, err = xorm.NewEngine(*driver, fmt.Sprintf("%s:%s@tcp(%s)/%s?charset=utf8", *usr, *pwd, *host, *dbname))
	if err != nil {
		fmt.Printf("error is: %v", err)
	}
if err = engine.Sync(new(models.User)); err != nil {
		fmt.Println("Database struct sync failed")
		fmt.Println("Engine.Sync ERRORS:", err)
	}
type XormController struct {
	Engine      *xorm.Engine
	XormSession *xorm.Session
}

func (c *XormController) Initially(http.ResponseWriter, *http.Request) bool {
	fmt.Println("This is xormsession Initially")
	//c.Engine = engine // the line result in the painc,why?

	return true
}

func (c *XormController) Before() http.Handler {
	c.Engine = engine
	if c.XormSession == nil {
		c.XormSession = c.Engine.NewSession()
	}
	fmt.Println("This is xormsession before")
	if err := c.XormSession.Begin(); err != nil {
		return nil
	}
	return nil
}

func (c *XormController) After() http.Handler {
	if c.XormSession != nil {
		c.XormSession.Commit()
		c.XormSession.Close()
		c.XormSession = nil
	}
	fmt.Println("This is xormsession after")
	return nil
}

// Finally commits the transaction and rolls it back if neither commit nor
// rollback were called.
func (c *XormController) Finally(http.ResponseWriter, *http.Request) bool {
	fmt.Println("This is xormsession Finally")
	if c.XormSession != nil {
		c.XormSession.Rollback()
		c.XormSession.Close()
		c.XormSession = nil
		return false
	}
	c.Engine = nil
	return true
}

```

**Comment by alkchr** [comment]
@xpbliss There was a change a few months ago: https://github.com/colegion/goal/issues/8#issuecomment-145337565. This hasn't been reflected in the docs yet. `Initially` and `Finally` must take 3 arguments:
```go
func (c *Controller) Initially(http.ResponseWriter, *http.Request, []string) bool
```
In your case there're only 2: `Finally(http.ResponseWriter, *http.Request) bool`.

**Comment by xpbliss** [comment]
oK

**Comment by xpbliss** [comment]
```Execution of a requested action (if non-nil value is returned goto FIN)```,  because  under  mnay conditions, a requested action return a render() and it is not a nil, the After() is not executed?

**Comment by alkchr** [comment]
Right, `After` is only executed if neither `Before` nor action returned a result. If you want something to be guaranteed to be executed after every action, you should use `Finally`.

**Comment by xpbliss** [comment]
Could  you  explain  orally  the use case for  the  ``` Initially,Before,After,Finally```?

**Comment by alkchr** [comment]
* `Initially` was only introduced as a way to pass `Request`, `ResponseWriter`, and current controller / action name to the controller. E.g. [`requests.Request`](https://github.com/colegion/contrib/blob/master/controllers/requests/requests.go#L23) sets `Request` so you can later use it in your actions as `c.Request`. (1)

* `Before` is what you should use most of the time for *auth* and all other kinds of checks that must be done before actions.

* `After` was an addition to `Before`. I didn't know at that time whether there would be a use-case for it. For completeness it was added. (2)

* `Finally` is useful as it may be used for catching panics, closing connections, committing DB transactions and so forth.

---------------

* (1): I'm thinking about what you've proposed above. Maybe, the number of "magic actions" can be reduced by introducing special struct tags. For illustration:

```go
type MyController struct {
	AnyName *http.Request `bind:"request"`
	SmthElse string       `bind:"action"`
}
```

*So, we can remove `Initially(... arguments here ...)` and instead use the tags that will tell which fields must be initialized with what kind of value.*

* (2): Can be safely removed as well.

---------------

So, it may be worth doing the following:

1. Remove `Initially`.
2. Remove `After`.
3. Rename `Finally` to `After`, change its signature.

Then we'll have only 2 magic actions: `Before` and `After`. The latter will be started after every action no matter what.

**Comment by alkchr** [comment]
> It is more better for you to finish the "Autoform", and then it maybe is ready for application.

Yeah, right. I'm currently thinking over the implementation details.

**Comment by alkchr** [comment]
Closing for now. Feel free to reopen or open a new issue if you need to.
